cmd/restic/cmd_cleanup_packs.go

```diff
@@ -0,0 +1,330 @@
+package main
+
+import (
+	"context"
+	"sync"
+
+	"github.com/restic/restic/internal/crypto"
+	"github.com/restic/restic/internal/debug"
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/pack"
+	"github.com/restic/restic/internal/repository"
+	"github.com/restic/restic/internal/restic"
+
+	"github.com/spf13/cobra"
+)
+
+var cmdCleanupPacks = &cobra.Command{
+	Use:   "cleanup-packs [flags]",
+	Short: "Remove packs not in index",
+	Long: `
+The "cleanup-packs" cleans up data in packs
+that is not referenced in any index files.
+
+When calling this command without flags, only packs
+that are completely unused are deleted.
+You can specify additional conditions to repack
+packs that are only partly used or too small.
+These packs will be downloaded and uploaded again which can be
+quite time-consuming for remote repositories.
+`,
+	DisableAutoGenTag: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runCleanupPacks(cleanupPacksOptions, globalOptions)
+	},
+}
+
+// CleanupIndexOptions collects all options for the cleanup-index command.
+type CleanupPacksOptions struct {
+	DryRun            bool
+	DataUnusedPercent int8
+	DataUnusedSpace   int64
+	DataUsedSpace     int64
+	TreeUnusedPercent int8
+	TreeUnusedSpace   int64
+	TreeUsedSpace     int64
+	RepackMixed       bool
+}
+
+var cleanupPacksOptions CleanupPacksOptions
+
+func init() {
+	cmdRoot.AddCommand(cmdCleanupPacks)
+
+	f := cmdCleanupPacks.Flags()
+	f.BoolVarP(&cleanupPacksOptions.DryRun, "dry-run", "n", false, "do not delete anything, just print what would be done")
+	f.Int8Var(&cleanupPacksOptions.DataUnusedPercent, "data-unused-percent", -1, "if > 0, repack data packs with >= given % unused space")
```

Comment:
The reason is that tree blobs are stored in the cache whilst data blobs are not. So if you expect that your cache already contains the tree packs which will be repacked, this will be very fast.
Hence there are use cases where it makes sense to differentiate.
And the added complexity is more flags and three extra condition lines in the switch that decides what to do with each packs...